shipyard_airflow/plugins/openstack_operators.py

```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import subprocess
+import sys
+import os
+import shlex
+
+from airflow.exceptions import AirflowException
+from airflow.models import BaseOperator
+from airflow.plugins_manager import AirflowPlugin
+from airflow.utils.decorators import apply_defaults
+
+class OpenStackOperator(BaseOperator):
+    """
+    Performs OpenStack CLI calls
+    :openrc_file: Path of the openrc file
+    :openstack_command: The OpenStack command to be executed
+    """
+    @apply_defaults
+    def __init__(self,
+                 openrc_file,
+                 openstack_command=None,
+                 xcom_push=False,
+                 *args, **kwargs):
+
+        super(OpenStackOperator, self).__init__(*args, **kwargs)
+        self.openrc_file = openrc_file
+        self.openstack_command = openstack_command
+        self.xcom_push_flag = xcom_push
+
+    def execute(self, context):
+        logging.info("Running OpenStack Command: " + self.openstack_command)
+
+        # Emulate "source" in bash. Sets up environment variables.
+        pipe = subprocess.Popen(". %s; env" % self.openrc_file, stdout=subprocess.PIPE, shell=True)
+        data = pipe.communicate()[0]
+        env = dict((line.split("=", 1) for line in data.splitlines()))
+        os.environ.update(env)
+
+
+        # Execute the OpenStack CLI Command
+        openstack_cli = subprocess.Popen(shlex.split(self.openstack_command), stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+
+        # Logs Output
+        logging.info("Output:")
+
+        line = ''
+        for line in iter(openstack_cli.stdout.readline, b''):
```

Comment:
Actually I had looked at the existing codes from the operators and followed their logic.  Hence I can see that the logs are shown in airflow UI under the log tab.  For instance, a sample output will look like the below:

```
[2017-06-23 03:11:24,350] {models.py:168} INFO - Filling up the DagBag from /usr/local/airflow/dags/openstack_api_call.py
[2017-06-23 03:11:24,945] {base_task_runner.py:112} INFO - Running: ['bash', '-c', u'airflow run openstack_cli service_list_task 2017-06-23T03:09:57.568664 --job_id 5 --raw -sd DAGS_FOLDER/openstack_api_call.py']
[2017-06-23 03:11:26,081] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:26,079] {__init__.py:57} INFO - Using executor CeleryExecutor
[2017-06-23 03:11:26,165] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:26,164] {driver.py:120} INFO - Generating grammar tables from /usr/lib/python2.7/lib2to3/Grammar.txt
[2017-06-23 03:11:26,208] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:26,204] {driver.py:120} INFO - Generating grammar tables from /usr/lib/python2.7/lib2to3/PatternGrammar.txt
[2017-06-23 03:11:26,468] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:26,467] {models.py:168} INFO - Filling up the DagBag from /usr/local/airflow/dags/openstack_api_call.py
[2017-06-23 03:11:26,987] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:26,987] {models.py:1128} INFO - Dependencies all met for <TaskInstance: openstack_cli.service_list_task 2017-06-23 03:09:57.568664 [queued]>
[2017-06-23 03:11:27,006] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:27,005] {models.py:1128} INFO - Dependencies all met for <TaskInstance: openstack_cli.service_list_task 2017-06-23 03:09:57.568664 [queued]>
[2017-06-23 03:11:27,006] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:27,005] {models.py:1328} INFO - 
[2017-06-23 03:11:27,006] {base_task_runner.py:95} INFO - Subtask: --------------------------------------------------------------------------------
[2017-06-23 03:11:27,007] {base_task_runner.py:95} INFO - Subtask: Starting attempt 1 of 2
[2017-06-23 03:11:27,007] {base_task_runner.py:95} INFO - Subtask: --------------------------------------------------------------------------------
[2017-06-23 03:11:27,007] {base_task_runner.py:95} INFO - Subtask: 
[2017-06-23 03:11:27,018] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:27,018] {models.py:1352} INFO - Executing <Task(OpenStackOperator): service_list_task> on 2017-06-23 03:09:57.568664
[2017-06-23 03:11:27,061] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:27,060] {openstack_operators.py:45} INFO - Running OpenStack Command: openstack service list
[2017-06-23 03:11:27,098] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:27,097] {openstack_operators.py:59} INFO - Output:
[2017-06-23 03:11:30,699] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,699] {openstack_operators.py:64} INFO - +----------------------------------+----------+----------+
[2017-06-23 03:11:30,700] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,699] {openstack_operators.py:64} INFO - | ID                               | Name     | Type     |
[2017-06-23 03:11:30,700] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,699] {openstack_operators.py:64} INFO - +----------------------------------+----------+----------+
[2017-06-23 03:11:30,700] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,699] {openstack_operators.py:64} INFO - | 3665ec03bb4045f6a33226923a0b283d | nova     | compute  |
[2017-06-23 03:11:30,700] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,699] {openstack_operators.py:64} INFO - | 6ef93b0dda6c45c2b95cd0302e72fe2a | glance   | image    |
[2017-06-23 03:11:30,701] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,699] {openstack_operators.py:64} INFO - | 7cf15d3215014d148c51c9b57ae85bb0 | keystone | identity |
[2017-06-23 03:11:30,702] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,700] {openstack_operators.py:64} INFO - | a20df35688944634965682c55f6135ea | neutron  | network  |
[2017-06-23 03:11:30,702] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,700] {openstack_operators.py:64} INFO - +----------------------------------+----------+----------+
[2017-06-23 03:11:30,749] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,747] {openstack_operators.py:70} INFO - Command exited with return code 0
[2017-06-23 03:11:34,979] {jobs.py:2107} INFO - Task exited with return code 0
```